File: 项目文件/backend/app/models/subscription.py
"""订阅模型"""

from __future__ import annotations

import uuid
from datetime import datetime
from enum import StrEnum
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, Enum, Float, ForeignKey, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base

if TYPE_CHECKING:
    from app.models.user import User


class BillingCycle(StrEnum):
    MONTHLY = "monthly"
    YEARLY = "yearly"


class SubscriptionStatus(StrEnum):
    ACTIVE = "active"
    CANCELLED = "cancelled"
    PAUSED = "paused"


class Subscription(Base):
    """订阅表"""

    __tablename__ = "subscription"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    provider: Mapped[str] = mapped_column(String(100), nullable=False)
    amount: Mapped[float] = mapped_column(Float, nullable=False)
    billing_cycle: Mapped[BillingCycle] = mapped_column(
        Enum(BillingCycle), default=BillingCycle.MONTHLY
    )
    next_billing: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    status: Mapped[SubscriptionStatus] = mapped_column(
        Enum(SubscriptionStatus), default=SubscriptionStatus.ACTIVE
    )
    owner_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("user.id"))
    created_at: Mapped[datetime] = mapped_column(
        DateTime, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, server_default=func.now(), onupdate=func.now()
    )

    owner: Mapped[User] = relationship("User", lazy="selectin")

File: 项目文件/frontend/src/types/subscription.ts
import type { Visibility } from '../utils/visibility';

export interface Subscription {
  id: string;
  name: string;
  provider: string;
  amount: number;
  billing_cycle: 'monthly' | 'yearly';
  next_billing: string | null;
  status: 'active' | 'cancelled' | 'paused';
  owner_id: string;
  visibility?: Visibility;
  restricted_users?: string[] | null;
  created_at: string;
  updated_at: string;
}

export interface SubscriptionCreate {
  name: string;
  provider: string;
  amount: number;
  billing_cycle?: string;
  next_billing?: string;
  visibility?: Visibility;
  restricted_users?: string[];
}

export interface SubscriptionUpdate {
  name?: string;
  provider?: string;
  amount?: number;
  billing_cycle?: string;
  next_billing?: string;
  status?: string;
  visibility?: Visibility;
  restricted_users?: string[];
}

export interface SubscriptionListResponse {
  items: Subscription[];
  total: number;
}
